src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -263,23 +265,60 @@ cudnnSetOpTensorDescriptor(opTensorDesc,
                     writer.indent--;
                     writer << "}\n";
                 }
-                else if ((arg0_shape.size() == 2) && (arg1_shape.size() == 2) &&
-                         (dot->get_reduction_axes_count() == 1))
+                //dot can be treat as matrix multiply
+                else
                 {
-                    // GEMM Call
-                    if (arg0_shape[0] != out[0].get_shape()[0] || // m
-                        arg1_shape[1] != out[0].get_shape()[1] || // n
-                        arg0_shape[1] != arg1_shape[0])           // k
+                    size_t reduction_axes = dot->get_reduction_axes_count();
+                    size_t m = 1, n = 1, k = 1;
+                    //check if input and output size correct
+                    for (size_t i = 0, idxa = arg0_shape.size() - 1; i < reduction_axes;
+                         i++, idxa--)
                     {
-                        throw std::runtime_error("input and output shape does not match for dot;");
+                        if (arg0_shape[idxa] != arg1_shape[i])
+                        {
+                            throw std::runtime_error(
+                                "input and output shape does not match for dot;");
+                        }
+                        else
+                        {
+                            k *= arg1_shape[i];
+                        }
                     }
+                    for (size_t i = 0; i < arg0_shape.size() - reduction_axes; i++)
+                    {
+                        if (arg0_shape[i] != out_shape[i])
+                        {
+                            throw std::runtime_error(
+                                "input and output shape does not match for dot;");
+                        }
+                        else
+                        {
+                            m *= arg0_shape[i];
+                        }
+                    }
+                    for (size_t i = 0, idxa = arg1_shape.size() - 1, idxb = out_shape.size() - 1;
```

Comment:
The use of these multi-index for loops make this difficult to read. Since these loops are helpers (not performance heavy), I think a loss in efficiency for the sake of readability would be better.

Reply:
good point.

Reply:
With modern compilers there is no loss of efficiency. Always shoot for "normal" and "readable".